ethernet: Synopsys DesignWare MAC driver

npitre opened the pull request:
This is a driver for the Synopsys DesignWare MAC. It should work
with the "DesignWare Cores Ethernet Quality-of-Service" versions 4.x
and 5.x.

This driver uses a zero-copy strategy, meaning that the hardware
reads and writes data directly from/to packet fragment buffers
provided by the network subsystem without first copying the data into
a dedicated DMA bounce buffer.

Platform specific setup is necessary for the hardware to work.
Currently, only the STM32H7X series is implemented and tested.
While this part needs refinement, this driver performs better and uses
far less code space than the HAL-based alternative. So I'd like for
this driver to be merged early so to facilitate (read: encourage)
contributions from interested people.

Not yet implemented:

- MDIO (it is WIP, currently relying on default PHY config)
- PTP support
- VLAN support
- various hardware offloads (when available)

npitre committed:
sys: util: provide bitfield construction/extraction utility macros

Those are especially useful with hardware device registers.
They behave the same way as their Linux equivalent, with a much
simpler implementation for now.

Example:

 #define REG_FIELD_A  GENMASK(6, 0)
 #define REG_FIELD_B  BIT(7)
 #define REG_FIELD_C  GENMASK(15, 8)
 #define REG_FIELD_D  GENMASK(31, 16)

Get:
 a = FIELD_GET(REG_FIELD_A, reg);
 b = FIELD_GET(REG_FIELD_B, reg);

Set:
 reg = FIELD_PREP(REG_FIELD_A, 1) |
       FIELD_PREP(REG_FIELD_B, 0) |
       FIELD_PREP(REG_FIELD_C, c) |
       FIELD_PREP(REG_FIELD_D, 0x40);

Modify:
 reg &= ~REG_FIELD_C;
 reg |= FIELD_PREP(REG_FIELD_C, c);

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
ethernet: Synopsys DesignWare MAC driver

This is a driver for the Synopsys DesignWare MAC. It should work
with the "DesignWare Cores Ethernet Quality-of-Service" versions 4.x
and 5.x.

This driver uses a zero-copy strategy, meaning that the hardware
reads and writes data directly from/to packet fragment buffers
provided by the network subsystem without first copying the data into
a dedicated DMA bounce buffer.

Platform specific setup is necessary for the hardware to work.
Currently, only the STM32H7X series is implemented and tested.
While this part needs refinement, this driver performs better and uses
far less code space than the HAL-based alternative.

Not yet implemented:

- MDIO (it is WIP, currently relying on default PHY config)
- PTP support
- VLAN support
- various hardware offloads (when available)

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

Nukersson commented on drivers/ethernet/eth_dwmac.c:
```diff
@@ -0,0 +1,805 @@
+	 * Semaphores are used to represent number of available descriptors.
+	 * The total is one less than ring size in order to always have
+	 * at least one inactive slot for the hardware tail pointer to
+	 * stop at, and to prevent our head indexes from looping back
```
```suggestion
	 * stop at and to prevent our head indexes from looping back
```

KozhinovAlexander reviewed (changes requested):
Thank you for your great job. I need more time fro proper review. I'll take a look later more precisely again.
It would be also really great to separate STM32H7 ´glue code´ from the driver itself even if additional *.c and *.h files would be needed.

npitre commented:
[ Sorry for not replying using the github comment tool as it is not blind friendly ]

Thank you for your initial review.

@Nukersson said:

> +/drivers/ethernet/*dwmac*                 @npitre
> 
> Should it be approved by you @nashif ?

This is for automatically being assigned as a reviewer for those files..
I'd be surprised if @nashif would be against me reviewing changes to my
own code from others. Or am I missing something?

> + * Copyright (c) 2021 BayLibre SAS
> 
> Correct copyright?

What's wrong with it?

> +static inline uint32_t phys_lo32(void *addr)
> +{
> +       /* the default 1:1 mapping is assumed */
> +       return (uintptr_t)addr;
> +}
> 
> ```suggestion
> #define PHYS_LO32(add)   ((uint32_t)(addr))
> ```
> 
> +static inline uint32_t phys_hi32(void *addr)
> +{
> +       /* the default 1:1 mapping is assumed */
> +       if (sizeof(void *) > 4) {
> +               /* trickery to avoid compiler warnings on 32-bit build targets */
> +               uint64_t hi = (uintptr_t)addr;
> +
> +               return hi >> 32;
> +       }
> +       return 0;
> +}
> 
> ```suggestion
> #define PHYS_HI32(addr)   ((uint32_t)arithmetic_shift_right(addr, 32)))
> ```

Here I disagree for those reasons:

- The 1:1 physical:virtual mapping on MMU system is assumed right now.
  But if that were not to be the case it would be much cleaner to do the
  proper translation in a function.

- `arithmetic_shift_right()` returns a signed value. Even if the cast will
  get rid of the sign bits in any case, sign-extending a memory address value
  just conceptually feels wrong to me.

- The compiler always inline and optimize those functions. Corresponding
  generated assembly is already optimal.

Otherwise I merged all the other suggestions.

npitre commented:
Latest push added:

- expanded the LSB macro name somewhat

- added more info to the Kconfig help text

- fixed descriptor physical address usage in the MMU case

Nukersson commented:
Just tested 8it with nucleo h745 zu q and M7 core. First compilation fails, cause KConfig does not disables default stm32 ETH hal driver. Please take care it and disable the current driver if yours is used and vice-versa.
P.S.: civetweb websocket server sample does not works with your driver. I will investigate it later.

npitre commented:
> Just tested it with nucleo h745 zu q and M7 core. First compilation
> fails, cause KConfig does not disables default stm32 ETH hal driver.
> Please take care it and disable the current driver if yours is used
> and vice-versa.

I disabled this driver when CONFIG_ETH_STM32_HAL is set.
I can do it the other way around, but not both as this would create
a Kconfig dependency loop.

> P.S.: civetweb websocket server sample does not works with your driver.
> I will investigate it later.

If I set:

```
CONFIG_NET_PKT_RX_COUNT=30
CONFIG_NET_PKT_TX_COUNT=30
CONFIG_NET_BUF_RX_COUNT=100
CONFIG_NET_BUF_TX_COUNT=100
```

then it works. Those are probably way higher than what's really necessary.
I didn't try to find out what the minimum values are.

tejlmand reviewed (commented):
> While this part needs refinement, this driver performs better and uses
far less code space than the HAL-based alternative. So I'd like for
this driver to be merged early so to facilitate (read: encourage)
contributions from interested people.

is this driver considered `EXPERIMENTAL` ?

npitre commented:
> is this driver considered `EXPERIMENTAL` ?

It is production ready for what it claims to do. It fulfills the needs 
of a product from $customer. It just doesn't do as much (like link state 
notification, VLAN and PTP) as the HAL-based alternative yet.

mbolivar-nordic commented:
>  Do you plan on generalizing to multiple instances?
> I personally have no such plan right now. More important features are still missing and begging for my time. Contributions from people who know how to do it are welcome though.

Knowing how to write a multi-instance driver should be considered a basic skill for Zephyr driver developers. There is a clear HOWTO in the documentation: https://docs.zephyrproject.org/latest/guides/dts/howtos.html#option-1-create-devices-using-instance-numbers
